src/components/EditorAndOutput.vue

```diff
@@ -441,23 +451,29 @@ function replaceOrAddGraph(filePath: FilePath, add: boolean) {
             <div v-if="sceneFile == null">Missing scene.json</div>
           </div>
           <CodeEditor
-            v-if="!props.visual"
+            v-if="openFile.editorType.value === 'shader'"
             class="self-stretch overflow-hidden flex-1"
             :keyed-code="openFile.code.value"
+            :is-readonly="openFile.isReadonly.value"
             :is-dark="store.isDark"
             @update="openFile.setNewCode($event)"
           >
           </CodeEditor>
           <CodeGraph
-            v-if="props.visual"
+            v-if="openFile.editorType.value === 'graph'"
+            :fs="props.fs"
+            :keyedGraph="openFile.code.value"
             :graphs="graphsDropdown"
-            @update="openFile.setNewCode($event)"
-            @save="
-              (fileName, content) => {
-                props.fs.writeTextFile(fileName, content);
+            @update="
+              (content) => {
+                if (openFile.path.value !== null) {
+                  saveGraphWgsl(openFile.path.value, content);
+                }
               }
             "
-            @load="(fileName, add) => replaceOrAddGraph(fileName, add)"
```

Comment:
The `@load` was replaced by just passing the filesystem to the graph. Please treat the filesystem there as a "readonly filesystem". :)